pallets/randomness/src/types.rs

```diff
@@ -0,0 +1,306 @@
+		};
+		if let Some(excess) = refundable_amount.checked_sub(&cost_of_execution) {
+			// refund cost_of_execution to caller of `fulfill`
+			T::ReserveCurrency::transfer(
+				&contract_address,
+				&T::AddressMapping::into_account_id(caller.clone()),
+				cost_of_execution,
+				KeepAlive,
+			)
+			.expect("just unreserved deposit + fee => cost_of_execution must be transferrable");
+			// refund excess to refund address
+			T::ReserveCurrency::transfer(
+				&contract_address,
+				&T::AddressMapping::into_account_id(self.refund_address),
+				excess,
+				KeepAlive,
+			)
+			.expect("just unreserved deposit + fee => excess must be transferrable");
+		} // TODO: else should log warning or emit event that no refund happened?
+	}
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct RequestState<T: Config> {
+	/// Underlying request
+	pub request: Request<T>,
+	/// Deposit taken for making request (stored in case config changes)
+	pub deposit: BalanceOf<T>,
+	/// All requests expire `T::ExpirationDelay` blocks after they are made
+	pub expires: T::BlockNumber,
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+/// Data required to make the subcallback and finish fulfilling the request
+pub struct FulfillArgs<T: Config> {
+	/// Original request
+	pub request: Request<T>,
+	/// Deposit for request
+	pub deposit: BalanceOf<T>,
+	/// Randomness
+	pub randomness: [u8; 32],
+}
+
+impl<T: Config> RequestState<T> {
+	pub(crate) fn new(request: Request<T>) -> RequestState<T> {
+		let expires =
+			frame_system::Pallet::<T>::block_number().saturating_add(T::ExpirationDelay::get());
+		// TODO: check that request.info.when is before `expires` (how to do it with EpochIndex)
+		RequestState {
+			request,
+			deposit: T::Deposit::get(),
+			expires,
+		}
+	}
+	/// Returns Ok(FulfillArgs) if successful
+	/// This should be called before the callback
+	pub fn prepare_fulfill(&self) -> Result<FulfillArgs<T>, DispatchError> {
+		// get the randomness corresponding to the request
+		let randomness: T::Hash = self.request.get_randomness()?;
+		// compute random output using salt
+		let randomness = Pallet::<T>::concat_and_hash(randomness, self.request.salt);
+		// No event emitted until fulfillment is complete
+		Ok(FulfillArgs {
+			request: self.request.clone(),
+			deposit: self.deposit,
+			randomness,
+		})
+	}
+	pub fn increase_fee(
+		&mut self,
+		caller: &H160,
+		fee_increase: BalanceOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		ensure!(
+			caller == &self.request.contract_address,
+			Error::<T>::OnlyRequesterCanIncreaseFee
+		);
+		let new_fee = self
+			.request
+			.fee
+			.checked_add(&fee_increase)
+			.ok_or(Error::<T>::RequestFeeOverflowed)?;
+		T::ReserveCurrency::reserve(
+			&T::AddressMapping::into_account_id(caller.clone()),
+			fee_increase,
+		)?;
+		self.request.fee = new_fee;
+		Ok(new_fee)
```

Comment:
Still being paranoid... here we calculate our new fee and call `reserve()` but we don't directly update storage (with our new fee value) -- it's up the caller I guess to do that. Ideally these can all happen atomically in one place